### PR TITLE
[RN] Status indicators

### PR DESCRIPTION
**Pull request by zbettenbuk**
Initial commit covering:
- Generic implementation of CircularLabel that web uses, reusing as much code as possible
- Audio only indicator following web's way of indicators

Please note, further status indicators to come in this PR, hence the general title

### CODE REVIEW

File: react/features/base/label/components/AbstractCircularLabel.js [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+    /**
+     * String that will be rendered as the label itself.
+     */
+    label: string
+};
+
+/**
+ * Abstract class for the {@code CircularLabel} component.
+ */
+export default class AbstractCircularLabel<P: Props> extends Component<P> {
```
Comment (virtuacoplenny) [inline react/features/base/label/components/AbstractCircularLabel.js]: Do you think it's still good to have this abstract if there isn't any abstract/base methods? I suppose the shared props are good but I wonder if it's enough.

File: react/features/base/label/components/styles.js [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+ */
+export default createStyleSheet({
+
+    /**
+     * The outermost view.
+     */
+    indicatorContainer: {
+        alignItems: 'center',
+        backgroundColor: '#808080',
+        borderRadius: 20,
```
Comment (virtuacoplenny) [inline react/features/base/label/components/styles.js]: Why not use a percentage? That way the width can change and maybe this value doesn't. I'm making large assumptions about how border-radius works though.

File: react/features/conference/components/styles.js [hunk 3]
```diff
@@ -35,5 +35,11 @@ export default createStyleSheet({
         // On iPhone X there is the notch. In the two cases BoxModel.margin is
         // not enough.
         top: BoxModel.margin * 3
+    },
+
+    videoQualityLabel: {
+        right: 0,
+        top: 0,
+        position: 'absolute'
```
Comment (virtuacoplenny) [inline react/features/conference/components/styles.js]: Does the label display at the very edge of the screen or is there margin/padding somewhere that brings it away from the edge?

File: react/features/base/label/components/styles.js (hunk 2 above)
Comment (zbettenbuk) [reply react/features/base/label/components/styles.js]: Unfortunately native doesn't support percentage :(

https://facebook.github.io/react-native/docs/view-style-props.html#borderradius

File: react/features/base/label/components/AbstractCircularLabel.js (hunk 1 above)
Comment (zbettenbuk) [reply react/features/base/label/components/AbstractCircularLabel.js]: Very good point, actually I was thinking about the same. Then I realised that when working with a real object oriented language, majority of the implemented inheritance is about property inheritance and there is a much fewer cases when you inherit behaviour. I like inheritance though, that's why I went with inheritance.

Taking the point further however (and this is now not about this particular change or comment), Facebook discourages anyone using inheritance in react at all. Which would be a good reason not to do it, we however already built our app over inheritance and code sharing over abstract components, so it would be weird to stop using it now. Maybe would be good to have a longer discussion about it though, but I think I would still be on the inheritance side :)

File: react/features/base/label/components/styles.js [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+        borderWidth: 0,
+        flex: 0,
+        height: 40,
+        justifyContent: 'center',
+        margin: BoxModel.margin,
+        opacity: 0.7,
+        width: 40
```
[the remark this answers is not in the document]
Comment (zbettenbuk) [reply react/features/base/label/components/styles.js]: Unfortunately the mobile style guide is much less detailed as the web one :( I tried several sizes, this was the one that I liked the most. Here is a screenshot how it looks now (iPhone 6), what do you think?

image

File: react/features/conference/components/styles.js (hunk 3 above)
Comment (zbettenbuk) [reply react/features/conference/components/styles.js]: In react native, 'absolute' is not really absolute. There is a significant margin that is added by the parent Conference element (see screenshot above).

File: react/features/video-quality/components/VideoQualityLabel.native.js [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+ * @private
+ * @returns {{
+ *     _style: Object
+ * }}
+ */
+function _mapStateToProps(state: Object) {
+    const _abstractProps = _abstractMapStateToProps(state);
+
+    const _style
+        = _abstractProps._audioOnly ? styles.indicatorAudioOnly : undefined;
```
[the remark this answers is not in the document]
Comment (zbettenbuk) [reply react/features/video-quality/components/VideoQualityLabel.native.js]: I was following the discussion we had earlier about moving most of the calculation into the external function to avoid re-render, but I just realised that it doesn't make sense here as this property is not prone to frequent change, so I'll move it in, thanks!

File: react/features/base/label/components/styles.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (virtuacoplenny) [reply react/features/base/label/components/styles.js]: Why green? Was it decided that green is the color to indicate audio only? Looks okay otherwise.

[the remark this answers is not in the document]
Comment (zbettenbuk) [reply react/features/base/label/components/styles.js]: I just went with the design mockups. Had to actually pick the color manually from the images because there was now hex code

File: react/features/base/label/components/AbstractCircularLabel.js [hunk 6]
```diff
@@ -0,0 +1,17 @@
+// @flow
+import { Component } from 'react';
+
+export type Props = {
+
+    /**
+     * String that will be rendered as the label itself.
+     */
+    label: string
+};
```
[the remark this answers is not in the document]
Comment (zbettenbuk) [reply react/features/base/label/components/AbstractCircularLabel.js]: You know I don't like to implement things that are not used ;)

File: react/features/video-quality/components/VideoQualityLabel.native.js [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+    /**
+     * Implements React {@link Component}'s render.
+     *
+     * @inheritdoc
+     */
+    render() {
+        const { _audioOnly, style, t } = this.props;
+
+        if (!_audioOnly) {
+            // We don't have info about the quality so no need for the indicator
```
[the remark this answers is not in the document]
Comment (zbettenbuk) [reply react/features/video-quality/components/VideoQualityLabel.native.js]: Actually it turned out that we don't need those indicators as they are not on the designs. I just left this comment here to make sure that if we ever decide to implement it, it should be here and not in a new component (also explains a bit why it's called VideoQualityLabel and not AudioOnly label)

About the video muted mode, good point, but I need confirmation for that too. I forward it to the hive mind as you say

File: react/features/video-quality/components/VideoQualityLabel.native.js [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+            return null;
+        }
+
+        return (
+            <CircularLabel
+                label = { t('videoStatus.audioOnly') }
```
[the remark this answers is not in the document]
Comment (zbettenbuk) [reply react/features/video-quality/components/VideoQualityLabel.native.js]: Same, not used code :/

About the video muted mode, good point, but I need confirmation for that too. I forward it to the hive mind as you say

File: react/features/recording/components/AbstractRecordingLabel.js [hunk 9]
```diff
@@ -0,0 +1,65 @@
+// @flow
+
+import { Component } from 'react';
+
+import { JitsiRecordingConstants } from '../../base/lib-jitsi-meet';
+
+/**
+ * NOTE: Web currently renders multiple indicators if multiple recording
```
[the remark this answers is not in the document]
Comment (zbettenbuk) [reply react/features/recording/components/AbstractRecordingLabel.js]: Yep, talking to Lenny in a bit

File: react/features/recording/components/RecordingLabel.native.js [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+        const { _display, mode, style, t } = this.props;
+
+        if (!_display) {
+            return null;
+        }
+
+        let labelKey;
+        let indicatorStyle;
+
+        switch (mode) {
```
[the remark this answers is not in the document]
Comment (zbettenbuk) [reply react/features/recording/components/RecordingLabel.native.js]: The default is undefined as the variables are there above. Or do you mean something different?

[the remark this answers is not in the document]
Comment (zbettenbuk) [reply react/features/recording/components/RecordingLabel.native.js]: Oh, I think I just understood what you meant. Okay, will change it, thanks!

File: react/features/recording/middleware.js [hunk 11]
```diff
@@ -0,0 +1,39 @@
...
+
+                if (recorderSession && recorderSession.getID()) {
+                    dispatch(
+                        updateRecordingSessionData(recorderSession));
+
+                    return;
+                }
+            });
+
+        break;
```
[the remark this answers is not in the document]
Comment (virtuacoplenny) [reply react/features/recording/middleware.js]: Nope, it's not. I should read everything before barfing out thoughts.

File: react/features/recording/components/AbstractRecordingLabel.js (hunk 9 above)
[the remark this answers is not in the document]
Comment (virtuacoplenny) [reply react/features/recording/components/AbstractRecordingLabel.js]: Third reply in a row... thought about this more and changing web actually seems better because web is trying to account for a situation that doesn't exist yet and hasn't been thought through UX-wise, which is supporting multiple recording/live streaming services at once (like twitch and youtube and whatever else). If a platform is to change it'd be web. I'd leave it to you to decide whether you want to do it or not, but you can dump a ticket on me to do it.

File: react/features/base/label/components/AbstractCircularLabel.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (saghul) [reply react/features/base/label/components/AbstractCircularLabel.js]: But it already is. On native. You just have a default value for it.

**Comment by saghul** [comment]
Looks like this needs a rebase.

File: react/features/video-quality/components/VideoQualityLabel.native.js (hunk 7 above)
[the remark this answers is not in the document]
Comment (saghul) [reply react/features/video-quality/components/VideoQualityLabel.native.js]: Let's please ask before making any assumptions. (FTR  hate the quality indicator with a passion, and don't want to see it on mobile). If we are not going to have them then this should be the AudioOnlyIndicator instead, don't you think?

File: react/features/video-quality/components/VideoQualityLabel.native.js (hunk 8 above)
[the remark this answers is not in the document]
Comment (zbettenbuk) [reply react/features/video-quality/components/VideoQualityLabel.native.js]: Okay, I haven't got answer for this, so let's leave it as it is now and when Nicole is back, we'll have an answer and change the indicator if need be

File: react/features/video-quality/components/VideoQualityLabel.native.js (hunk 7 above)
[the remark this answers is not in the document]
Comment (zbettenbuk) [reply react/features/video-quality/components/VideoQualityLabel.native.js]: I really don't want to change it as it's in line with web now (has an abstract class, shares some code...etc) and eventually that's our goal.

On the other hand, based on wording and UX on web, audio-only is a actually a level of video quality (you set it in the video quality menu).

File: react/features/base/label/components/AbstractCircularLabel.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (zbettenbuk) [reply react/features/base/label/components/AbstractCircularLabel.js]: I don't really get what you mean. In order to make the size adjustable (which is not needed atm) we need to extract the size and border setting into a dynamic style object generated run-time (on render, I guess), and that would be a code that we wouldn't use because we'd only go with the default size for now. Or did I misunderstand you?

[the remark this answers is not in the document]
Comment (saghul) [reply react/features/base/label/components/AbstractCircularLabel.js]: No, that's what I meant, yes. 🤷‍♂️

File: react/features/video-quality/components/VideoQualityLabel.web.js [hunk 12]
```diff
@@ -1,22 +1,46 @@
...
 /**
  * A map of video resolution (number) to translation key.
  *
  * @type {Object}
  */
 const RESOLUTION_TO_TRANSLATION_KEY = {
-    720: 'videoStatus.hd',
-    360: 'videoStatus.sd',
-    180: 'videoStatus.ld'
+    '720': 'videoStatus.hd',
```
Comment (saghul) [inline react/features/video-quality/components/VideoQualityLabel.web.js]: why the change?

also, come tho think of it, the computation for RESOLUTIONS below can be simplified.

Comment (zbettenbuk) [reply react/features/video-quality/components/VideoQualityLabel.web.js]: This file was not flow-ified before and now that I added the annotation, it required me to change it.

I agree with simplification, but I think this is not part of this PR - Although I'm working on a larger refactor for this component and I'll include that simplification into that PR for sure

Comment (saghul) [reply react/features/video-quality/components/VideoQualityLabel.web.js]: It's one line though.

File: react/features/conference/components/Indicators.native.js [hunk 13]
```diff
@@ -0,0 +1,78 @@
...
+    const { visible } = state['features/filmstrip'];
+
+    return {
+        /**
+         * The indicator which determines whether the filmstrip is visible.
+         *
+         * @private
+         * @type {boolean}
+         */
+        _filmstripVisible: visible && participantCount > 1
```
Comment (saghul) [inline react/features/conference/components/Indicators.native.js]: why the check for participantcount? also, please use the new helper utility for gettting the participants count, it's more future-proof.

File: react/features/conference/components/Conference.native.js [hunk 14]
```diff
@@ -10,18 +10,16 @@ import { appNavigate } from '../../app';
...
 import { TestConnectionInfo } from '../../base/testing';
 import { createDesiredLocalTracks } from '../../base/tracks';
 import { ConferenceNotification } from '../../calendar-sync';
 import { Filmstrip } from '../../filmstrip';
 import { LargeVideo } from '../../large-video';
-import { RecordingLabel } from '../../recording';
 import { setToolboxVisible, Toolbox } from '../../toolbox';
-import { VideoQualityLabel } from '../../video-quality';
 
+import Indicators from './Indicators';
```
Comment (saghul) [inline react/features/conference/components/Conference.native.js]: Can we find a better name for this?

File: react/features/recording/middleware.js [hunk 15]
```diff
@@ -0,0 +1,39 @@
...
+ */
+MiddlewareRegistry.register(({ dispatch }) => next => action => {
+    const result = next(action);
+
+    switch (action.type) {
+    case CONFERENCE_WILL_JOIN: {
+        const { conference } = action;
+
+        conference.on(
+            JitsiConferenceEvents.RECORDER_STATE_CHANGED,
```
Comment (saghul) [inline react/features/recording/middleware.js]: I think this belongs here: https://github.com/jitsi/jitsi-meet/blob/master/react/features/base/conference/actions.js#L69

Comment (zbettenbuk) [reply react/features/recording/middleware.js]: I started to move it to what you suggested and then I realised that it would mean that a base feature would reference a non-base feature which I generally don't like. We also talked about this with Lyubo (since we just had our catchup while I was working on it) and he also said that recording can easily remain a black box and keep its event handling in there. What do you think?

File: react/features/conference/components/Indicators.native.js (hunk 13 above)
Comment (zbettenbuk) [reply react/features/conference/components/Indicators.native.js]: Sure, I change the function, thanks! But we need to check the participant counts, otherwise it won't properly work (see Filmstrip code)

File: react/features/conference/components/Conference.native.js (hunk 14 above)
Comment (zbettenbuk) [reply react/features/conference/components/Conference.native.js]: I'll try to come up with some fancier ;)

File: react/features/conference/components/Indicators.native.js (hunk 13 above)
Comment (zbettenbuk) [reply react/features/conference/components/Indicators.native.js]: Checked that function and I'm afraid that won't work for us here. That function excludes the bot participants, that the Filmstrip doesn't and we need to behave exactly like the Filmstrip because thats position is the thing that what we want to address here.

File: react/features/video-quality/components/VideoQualityLabel.web.js (hunk 12 above)
Comment (zbettenbuk) [reply react/features/video-quality/components/VideoQualityLabel.web.js]: Ok, shoot!

**Comment by saghul** [comment]
Jenkins please test this please.

**Comment by saghul** [comment]
(Merged with an unrelated test failure)
